geomstats/geometry/pullback_metric.py

```diff
@@ -86,81 +51,48 @@ def metric_matrix(self, base_point=None, n_jobs=1, **joblib_kwargs):
-        basis_elements = gs.eye(self.dim)
-
-        @joblib.delayed
-        @joblib.wrap_non_picklable_objects
-        def pickable_inner_product(i, j):
-            immersed_basis_element_i = gs.squeeze(
-                gs.matvec(jacobian_immersion, basis_elements[i])
-            )
-            immersed_basis_element_j = gs.squeeze(
-                gs.matvec(jacobian_immersion, basis_elements[j])
-            )
-            return self.embedding_metric.inner_product(
-                immersed_basis_element_i,
-                immersed_basis_element_j,
-                base_point=immersed_base_point,
-            )
+        dim, dim_embedding = self._space.dim, self._space.embedding_space.dim
+        is_vec = check_is_batch(self._space, base_point)
 
-        pool = joblib.Parallel(n_jobs=n_jobs, **joblib_kwargs)
-        out = pool(
-            pickable_inner_product(i, j)
-            for i, j in itertools.product(range(self.dim), range(self.dim))
-        )
-        metric_mat = gs.reshape(gs.array(out), (-1, self.dim, self.dim))
-        return metric_mat[0] if base_point.ndim == 1 else metric_mat
+        immersed_base_point = self._space.immersion(base_point)
+        jacobian_immersion = self._space.jacobian_immersion(base_point)
 
-    def _hessian_immersion_func(self):
-        """Compute the Hessian of the immersion.
+        basis_elements = gs.eye(dim)
 
-        Returns
-        -------
-        hessian_immersion : list of callable
-            List of embedding_dim hessians of the scalar
-            functions defining the components of the immersion.
-        """
-
-        def _immersion(base_point, a):
-            """Compute the component a of the immersion at a base point.
-
-            Parameters
-            ----------
-            base_point : array-like, shape=[..., dim]
-                Base point.
-            a : int
-                Index of the component of the immersion.
-            """
-            return gs.array([self.immersion(base_point)[a]])
-
-        hessians = [
-            gs.autodiff.hessian(partial(_immersion, a=a))
-            for a in range(self.embedding_metric.dim)
-        ]
-        return hessians
+        if is_vec:
+            reshaped_jacobian_immersion = gs.reshape(jacobian_immersion, (-1, dim))
+            reshaped_immersed_basis_elements = gs.matvec(
+                reshaped_jacobian_immersion, basis_elements
+            )
+            immersed_basis_elements = gs.moveaxis(
+                gs.reshape(reshaped_immersed_basis_elements, (dim, -1, dim_embedding)),
+                0,
+                1,
+            )
 
-    def hessian_immersion(self, base_point):
-        """Compute the Hessian of the immersion.
+        else:
+            immersed_basis_elements = gs.matvec(jacobian_immersion, basis_elements)
 
-        Parameters
-        ----------
-        base_point : array-like, shape=[..., dim]
-            Base point.
+        elems = {}
```

Comment:
why a dictionary of tuples? why not some array?

Reply:
how about renaming elems and mat to something more specific / understandable for mathematicians?